### PR TITLE
const evaluator: array values and initialization

### PR DESCRIPTION
**Pull request by marcrasi**
Adds array values, initialization operations, and some tests that initialize arrays but that do not check the contents of the initialized arrays.

This PR includes support for AllocateUninitializedArray, but only in the flow-sensitive case. When I tried completely removing AllocateUninitializedArray (per your suggestion in https://github.com/apple/swift/pull/21703), I realized that it was no longer possible to write tests exercising the `RK_ArrayAddress` representation and the `getIndexedElement`/`setIndexedElement` changes. I think that those are important pieces of basic array support that should be included in the initial PR.

AllocateUninitializedArray in the flow-sensitive case exercises those features, so I added it back. It's still a lot simpler than https://github.com/apple/swift/pull/21703, because most of the AllocateUninitializedArray complexity was in the top level case.

### CODE REVIEW

File: stdlib/public/core/Array.swift [hunk 1]
```diff
@@ -803,6 +803,7 @@ extension Array: RangeReplaceableCollection {
   ///     // Prints "true"
   @inlinable
   @_semantics("array.init")
+  @_semantics("array.init_empty")
   public init() {
```
Comment (ravikandhadai) [inline stdlib/public/core/Array.swift]: I wonder if we need a new semantics attribute here or instead check that the parameters of the initializer are empty  and that the receiver object is Array, and return unknown if it is not.

@eeckstein This PR is adding support for Arrays to the interpreter. We would like to handle only `Array.init()`, and will have use a semantics annotation to hardcode its semantics in the interpreter. There seems to be an "array.init" semantics attribute on Array initializers, but it is shared by empty and non-empty initializes as well as by other types: `ArraySilce` and `ContiguousArray`. So, to distinguish `Array.init()` there are two options, either we add another semantics attribute to `Array.init()`, or check the receiver type and the parameter count in the interpreter to identify this initializer from others. Just wondering what approach you would recommend here?

File: stdlib/public/core/ArrayShared.swift [hunk 2]
```diff
@@ -27,6 +27,7 @@ public struct _DependenceToken {
 /// - Precondition: `storage` is `_ContiguousArrayStorage`.
 @inlinable // FIXME(inline-always)
 @inline(__always)
+@_semantics("array.allocate_uninitialized")
 public // COMPILER_INTRINSIC
```
Comment (ravikandhadai) [inline stdlib/public/core/ArrayShared.swift]: I think this being a compiler intrinsic and also something declared in `KnownDecls.def`, we should not have a semantics attribute here but instead just compare to `astContext.getAllocateUninitializedArray()`. 

@eeckstein It would be very helpful if you can share your thoughts on this issue as well.

**Comment by ravikandhadai** [comment]
@marcrasi Sorry for the delay. I just made another PR which when merged will make the changes here simpler.  Btw, I came across this function: `ArrayAllocation::mapInitializationStores` it seems very similar to the original `_handleAllocateUninitialized` or a similar helper function for handling top-level array literals

**Comment by marcrasi** [comment]
I rebased this to master and addressed all your non-`_semantics` comments! I'll rebase again and adjust the semantics stuff once your PR merges.

**Comment by ravikandhadai** [comment]
Great. The @_semantics PR just got merged.

**Comment by marcrasi** [comment]
Updated!

**Comment by ravikandhadai** [comment]
Cool. LGTM. @devincoughlin let me know if this looks good to you.

**Review by ravikandhadai — APPROVED** [review]

**Comment by ravikandhadai** [comment]
@swift-ci Please test

**Comment by marcrasi** [comment]
Ping on this. Are we waiting on a look for @devincoughlin ?

**Comment by ravikandhadai** [comment]
@marcrasi Yes

File: include/swift/SIL/SILConstants.h [hunk 3]
```diff
@@ -114,6 +114,12 @@ class SymbolicValue {
 
     /// This represents an index *into* a memory object.
     RK_DerivedAddress,
+
+    /// This represents an array value.
+    RK_Array,
+
+    /// This represents an address of a memory object containing an array.
+    RK_ArrayAddress,
```
Comment (ravikandhadai) [inline include/swift/SIL/SILConstants.h]: @marcrasi Sorry for the delay in the responding to this PR. Based on discussions with @devincoughlin, we have a few questions/comments about this PR. We are trying to understand the motivation for the two representations for an array: `RK_Array` and `RK_ArrayAddress`. Swift's representation of an Array, in principle, consists of `struct` which has a pointer to a backing storage that contains the elements. It seems to us that RK_Array (i.e, ArraySymbolicValue) models the backing storage which stores the elements.  RK_ArrayAddress is modeling the more accurate view of the array as a struct with a pointer to the backing storage namely RK_Array. It seems this distinction between an Array and its backing storage becomes important for modeling `_allocateUninitializedArray` function which returns them separately and initializes the backing storage directly through the pointer.  Our first question is, is our understanding right? 

If that is the case, it seems to us that RK_ArrayAddress is possibly the more accurate view of an array and RK_Array can be seen as  a more optimal representation when the distinction between the array and the backing storage can be safely discarded. Therefore, the name `RK_ArrayAddress` seems a bit misleading as it is not representing an address of an Array, but rather a more accurate view of an Array.  Would it be possible to give RK_Array, RK_ArrayAddress, ArraySymbolicValue another name and also add some documentation? For instance, what do you think about, `SymbolicArrayBuffer` for ArraySymbolicValue, `RK_ArrayBuffer` for RK_Array and `RK_ArrayStruct` for RK_ArrayAddress.   

Ideally, I think we can create a new symbolic value for Array that will manage the differences in the representation. But, I think I can do that in another PR, after this lands.

File: lib/SIL/SILConstants.cpp [hunk 4]
```diff
@@ -177,6 +201,16 @@ SymbolicValue::cloneInto(ASTContext &astContext) const {
...
+  case RK_ArrayAddress: {
+    CanType elementType;
+    auto elts = getArrayValue(elementType);
+    SmallVector<SymbolicValue, 4> results;
+    results.reserve(elts.size());
+    for (auto elt : elts)
+      results.push_back(elt.cloneInto(astContext));
+    return getArray(results, elementType, astContext);
+  }
   }
```
Comment (ravikandhadai) [inline lib/SIL/SILConstants.cpp]: It seems clone is collapsing the representation of RK_Array and RK_ArrayAddress. It ignores the differences. Is this intended?

File: include/swift/SIL/SILConstants.h (hunk 3 above)
Comment (ravikandhadai) [reply include/swift/SIL/SILConstants.h]: I think `SymbolicArrayStorage` and `RK_ArrayStorage` are better alternative names for `SymbolicArrayBuffer` and `RK_ArrayBuffer` I proposed above. This is inline with `_ContiguousArrayStorage` class used in by struct Array.

Comment (marcrasi) [reply include/swift/SIL/SILConstants.h]: I like these name suggestions, thanks! Your comments and name suggestions are making it clear to me that these RKs are two fundamentally different things that should actually have different `Kind`s.

Also, I'm pretty sure that the place below where you noticed the collapsing is a bug. (Probably not exposed by anything yet, but likely to become exposed when there are more array features.)

Finally, now that it's clear that RK_ArrayAddress is just a struct containing a pointer, why not make it a RK_Aggregate, like all the normal structs? This should simplify things!

So in summary, the total changes from the current state of this PR are:
* `RK_Array` => `RK_ArrayStorage`
* `ArraySymbolicValue` => `SymbolicArrayStorage`
* `RK_ArrayAddress` => eliminated, replaced by the existing `RK_Aggregate`

I haven't tried this out yet so there may be some difficulty I'm not seeing. So I'll try it out soon and let you know how it goes.

Comment (ravikandhadai) [reply include/swift/SIL/SILConstants.h]: Sounds great. :-)

Comment (ravikandhadai) [reply include/swift/SIL/SILConstants.h]: On reflection, I think that it may be a good idea to have a special kind for "Array" aggregates (in addition to `ArrayStorage` kind) and not treat them as general aggregates. In fact, I even think we should have special `ArraySymbolicValue` (in addition to `SymbolicArrayStorage`). The following are some reasons for that:

* We don't want to make it appear that the interpreter supports arbitrary aggregates with pointers.  Having a special "kind" for such an aggregate, makes this clear. 

* `get/setIndexedElement` would need to handle `addresses` (which could now appear inside aggregates). Normally, when there are addresses, `setIndexedElement` would be expected to update the underlying memory object in place, as we want everything that aliases with the pointer to also see the changes when `setIndexedElement` completes. But, arrays are COW and provide a value semantics.  This means we have to special case the `address` case for arrays, like creating a new memory object and not in-place updating the underlying memory object. This would appear confusing, as it is only correct for a type like Array that uses pointers underneath but provides a value semantics. 

To address both of this, we may rather have a new symbolic value for "array" which just stores a single `DirectAddressValue` or even just a `SymbolicMemoryObject` containing a `SymbolicArrayStorage`.   This `ArraySymbolicValue` may support operations like `getAddressOfStorage()` which creates a direct address from the underlying SymbolicMemoryObject, and `getStorage()` which returns the storage. `setIndexedElement` can just retrieve and (immutably) update the array storage and create a new ArraySymbolicValue with the new storage and return it.  This design essentially means that the only way to in-place update an `ArraySymbolicValue` instance  is to use `getAddressOfStorage` and update the storage through the address. This is also the *only way* AFAIK in Swift also to mutably update an array storage in place, and is done only to initialize the array.

**Comment by ravikandhadai** [comment]
This PR: https://github.com/apple/swift/pull/27344/ has addressed the comments posted here and has been merged.
